### ISSUE TITLE
Export blocks in diff-friendlier format

### ISSUE THREAD

**Issue by kephas** [jmoenig/Snap--Build-Your-Own-Blocks#2144]
The current export for blocks creates a single line, which makes it almost impossible to use diff and version control on it.

**Comment by brianharvey** [comment]
+1.  Hard to read, hard to edit.  We'll get to it at some point.

**Comment by jguille2** [comment]
Yes, but this is a problem with all our xml files (libs and examples).

Maybe we can try to use always well-formatted xmls. But this must be revised globally. For example, we will have problems with some elements that are not following xml rules, like menu options or custom block translations, separated by break lines.

**Comment by brianharvey** [comment]
We don't have to do a beautiful. well-indented formatting.  All that's needed is to include newlines after a few selected tags, e.g., </block> and </sprite>.  Or, even easier, a newline after _every_ closing tag.

**Comment by kephas** [comment]
I agree, the goal should not, at least at first, to make it pretty and readable, but at list diffable. That would already be a huge help.

**Comment by jmoenig** [comment]
Line based diffing is not something I'm much interested in. With blocks-based computing we want something that compares AST-nodes, not lines of code.

**Comment by kephas** [comment]
Right now, I'm writing a drawing library in Snap! to help the children I work with make complicated procedural drawings. Doing a matrix linear transformation with trigonometry goes a bit over the head of the typical 11 years old, so I'm providing the releavant blocks. :-)

The thing is, I would like to cooperate with other educators and programmers in developing that library, but that would mean using Git to merge our code. That is where line based diffing of Snap! code comes into play.

If you plan to write a block-based diff program that I can plug into Git, that would be incredibly awesome, I agree. In the meantime, it would be great if we can use the existing tools.

**Comment by jmoenig** [comment]
Yep, I agree it would be feasible first step and might even make sense. But there are other issues with working with serialized data. One of them is that the order of elements is not always the same when Snap serializes elements, and that many of them - e.g. scripts - aren't even ID'd. But yeah, maybe we should do it anyway. Of course, I'm really, really, interested in being able to diff and merge things in tools inside Snap! in the medium run

**Comment by brianharvey** [comment]
@jmoenig I don't think that your point about AST diffing is unique to blocks languages; text languages have syntax trees, too, and good IDEs provide syntax-aware editing tools.  (Of course, really lucky programmers get to use a language in which the language text _is_ the AST!)

Meanwhile we could make an easy improvement that would help quite a bit.  The only cost I can see is that project files would be slightly bigger.  But really only slightly, since most of the weight of a project file is costumes.  Statistics:

count-change:  32280 bytes, 488 closing tags
vee: 20102 bytes, 182 closing tags

(These numbers of closing tags are measured by looking for the string "&lt;/" so they are slight overestimates if that two-byte sequence appears in a picture.)

**Comment by cycomachead** [comment]
> (These numbers of closing tags are measured by looking for the string "</" so they are slight overestimates if that two-byte sequence appears in a picture.)

It won't! ;) `<` isn't in the bas64 dictionary. 

I also think pretty printing is useful, but new lines would work well enough and are fairly minimal effort. (And don't incur the size penalty that a load of properly indented spaces would.)
